### PR TITLE
feat(Surveys): Allow actions to activate surveys

### PR DESCRIPTION
**Pull request by Phanatic**
## Description
This PR allows us to activate surveys based on user actions performed in the browser.

### Not supported
1. Matching person properties in Actions,  when matching surveys. We will not implement this as we don't want to download person properties on the client for surveys.
2. Matching event properties in Actions, when matching surveys. We could do this in a follow-up PR because the event properties are available on the client-side.

## Changes

Here are the set of changes I needed to make to implement action matching in posthog-js

### AutoCapture
1. Autocapture can now register a set of html element selectors. 
4. These html selectors, are checked against the target element **that** was captured
5. If there is a match, those element selectors are sent along as `$element_selectors` property for a given event.

### Survey Event Receiver
1. Instead of maintaining a map of Survey-ID to its actions or events, we maintain a map of Action/Event to the surveys that are activated by it.
1. We now look for, and register any surveys that are activated by events with the `ActionMatcher`.
2. Changed `on` to be `onEvent` and added a new `onAction` event handler.
3. `onAction` listens on the `actionCaptured` event in `ActionMatcher` described below.

### ActionMatcher
This is a pared down version of the action-matcher found in the ingestion pipeline, main differences are : 
1. Uses the `$element_selectors` property introduced in AutoCapture to match events to an action step.
2. Uses the `elementsJson` that is passed along as part of CaptureResult to check other properties like tagname, text, url etc
4. Raises an `actionCaptured` event when a `CaptureResult` matches a known action step.

### Flowchart for matching actions on the client-side.
``` mermaid
flowchart TD
    A[User clicks button] -->|AutoCapture| B(Analyze element)
    B --> |Element matches known selector|C(Set $element_selectors property)
    C -->D[Capture Action]
    B --> |Element does not match known selector|D[Capture Action]
    D --> |Raise eventCaptured with CaptureResult|E[Event Handlers]
    E --> F[Survey Event Receiver onEvent] 
    F --> H[Match event name to known event names]
    H --> |eventName matches a survey's events|I[Add survey to activatedSurveys]
    E --> G[Action Matcher]
    G --> |Check Action Step Matches CaptureResult|J[Raise actionCaptured with actionName]
    J --> K[Survey Event Receiver onAction]
    K --> L[Match action name to known action names]
    L --> |actionName matches a survey's actions|M[Add survey to activatedSurveys]
```
...

## Checklist
- [ ] Remove all debugging code 😄 
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
     - [x] Survey Event Receiver
     - [ ] AutoCapture
     - [ ] Action Matcher
     - [ ] End-to-End Cypress tests
- [x] Accounted for the impact of any changes across different browsers
- [x] Accounted for backwards compatibility of any changes (no breaking changes in posthog-js!)

### CODE REVIEW

File: src/__tests__/utils/survey-event-receiver.test.ts [hunk 1]
```diff
@@ -1,157 +1,323 @@
...
+
+        it('receiver activates survey on event', () => {
+            const surveyEventReceiver = new SurveyEventReceiver(instance)
+            surveyEventReceiver.register(surveysWithEvents)
+            surveyEventReceiver.onEvent('billing_changed')
+            const activatedSurveys = surveyEventReceiver.getSurveys()
+            expect(activatedSurveys).toContain('first-survey')
+        })
+
+        it('receiver removes survey from list after its shown', () => {
```
Comment (jurajmajerik) [inline src/__tests__/utils/survey-event-receiver.test.ts:114]:
```suggestion
        it('receiver removes survey from list after it is shown', () => {
```

File: src/__tests__/utils/survey-event-receiver.test.ts [hunk 2]
```diff
@@ -1,157 +1,323 @@
...
+                uuid: '',
+                properties: {
+                    $survey_id: 'first-survey',
+                },
+            })
+
+            expect(surveyEventReceiver.getSurveys()).toEqual([])
+        })
+
+        it('receiver activates same survey on multiple event', () => {
```
Comment (jurajmajerik) [inline src/__tests__/utils/survey-event-receiver.test.ts:135]:
```suggestion
        it('receiver activates same survey on multiple events', () => {
```

File: src/__tests__/utils/survey-event-receiver.test.ts [hunk 3]
```diff
@@ -1,157 +1,323 @@
...
+                    },
+                },
+            } as unknown as Survey
+            autoCaptureSurvey.conditions.actions.values = [createAction(2, '$match_event_name')]
+            const surveyEventReceiver = new SurveyEventReceiver(instance)
+            console.log(`TEST:: registering surveys to observe`)
+            surveyEventReceiver.register([autoCaptureSurvey, myPageViewSurvey])
+            surveyEventReceiver._getActionMatcher().on('$match_event_name', createCaptureResult('$match_event_name'))
+            console.log(`TEST:: getting surveys to test`)
+            expect(surveyEventReceiver.getSurveys()).toEqual(['first-survey'])
```
Comment (jurajmajerik) [inline src/__tests__/utils/survey-event-receiver.test.ts:270]: For better readability, I'd prefer the id of `autoCaptureSurvey` to match the variable name, i.e., `auto-capture-survey`.

File: src/extensions/surveys/action-matcher.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        actions.forEach((action) => {
+            this.actionRegistry?.add(action)
+            action.steps?.forEach((step) => {
+                this.actionEvents?.add(step?.event || '')
+            })
+        })
+        // eslint-disable-next-line no-console
+        console.log(`in action-matcher eventNames known are `, this.actionEvents.values())
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:42]: remove?

File: src/extensions/surveys/action-matcher.ts [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        // }
+    }
+
+    on(eventName: string, eventPayload?: CaptureResult) {
+        if (eventPayload == null || eventName.length == 0) {
+            return
+        }
+
+        if (!this.actionEvents.has(eventName) && !this.actionEvents.has(<string>eventPayload?.event)) {
+            // throw new Error(`unknown event ${eventName}, I only know about : ${JSON.stringify(this.actionEvents.values())}`)
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:67]:
```suggestion

```

File: src/extensions/surveys/action-matcher.ts [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            return
+        }
+
+        if (this.actionRegistry && this.actionRegistry?.size > 0) {
+            this.actionRegistry.forEach((action) => {
+                if (this.checkAction(eventPayload, action)) {
+                    // // eslint-disable-next-line no-console
+                    // console.log(`in action-matcher emitting observed event for action`, action.name)
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:75]:
```suggestion
```

File: src/extensions/surveys/action-matcher.ts [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    private checkStep = (event?: CaptureResult, step?: ActionStepType): boolean => {
+        // eslint-disable-next-line no-console
+        console.log(
+            ` in checkStep,  event is [${event?.event}],  step.event is ${
+                step?.event
+            } checkStepEvent is [${this.checkStepEvent(event, step)}] checkStepUrl  is [${this.checkStepUrl(
+                event,
+                step
+            )}] checkStepElement is [${this.checkStepElement(event, step)}]`
+        )
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:113]: I assume you didn't want to leave any of these console.logs behind :)

I don't use eslint-disable to make sure the pre-commit check catches it.

```suggestion
```

File: src/extensions/surveys/action-matcher.ts [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        // Use a simple backslash escape when it’s always valid, and a `\xnn` escape when the simpler form would be disallowed by Unicode patterns’ stricter grammar.
+        return pattern.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&').replace(/-/g, '\\x2d')
+    }
+
+    private checkStepElement(event?: CaptureResult, step?: ActionStepType): boolean {
+        // CHECK CONDITIONS, OTHERWISE SKIPPED
+        if (step?.href || step?.tag_name || step?.text) {
+            const elements = this.getElementsList(event)
+            // eslint-disable-next-line no-console
+            console.log(` in action matcher, elements is `, elements)
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:169]:
```suggestion
```

File: src/extensions/surveys/action-matcher.ts [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+                !elements.some((element) => {
+                    // eslint-disable-next-line no-console
+                    console.log(
+                        `in action matcher, element is `,
+                        element,
+                        ` attributes are `,
+                        element.attributes,
+                        `element.text is `,
+                        element.text
+                    )
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:180]:
```suggestion
```

Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:180]:
```suggestion
```

File: src/extensions/surveys/action-matcher.ts [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        )
+
+        // eslint-disable-next-line no-console
+        console.log(
+            ` in action matcher, event is ${event?.event} step?.selector is [${
+                step?.selector
+            }, step?.selector !== event?.properties?.$element_selector is ${
+                step?.selector !== event?.properties?.$element_selector
+            } event?.properties?.$element_selector is ${event?.properties?.$element_selector}`
+        )
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:223]:
```suggestion
```

File: src/extensions/surveys/action-matcher.ts [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+            !event?.properties.$element_selectors?.contains(step?.selector)
+        ) {
+            // eslint-disable-next-line no-console
+            console.log(` in action matcher, event is ${event?.event} checkStepElement is returning false`)
+            return false // SELECTOR IS A MISMATCH
+        }
+
+        // if (step?.selector && !this.checkElementsAgainstSelector(event, step?.selector)) {
+        //     return false // SELECTOR IS A MISMATCH
+        // }
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:236]:
```suggestion
```

File: src/extensions/surveys/action-matcher.ts [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        return true
+    }
+
+    private getElementsList(event?: CaptureResult): SurveyElement[] {
+        if (event?.properties.$elements == null) {
+            return []
+        }
+
+        // eslint-disable-next-line no-console
+        console.log(` in action matcher: getElementsList, elements is `, event?.properties.$elements)
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:246]:
```suggestion
```

File: src/extensions/surveys/action-matcher.ts [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        return event?.properties.$elements as unknown as SurveyElement[]
+    }
+
+    // private mutateCaptureResultWithElementsList(event: CaptureResult: Element[] {
+    //      event.elementsList = event.elementsList.map((element) => ({
+    //     ...element,
+    //     attr_class: element.attributes?.attr__class ?? element.attr_class,
+    //     $el_text: element.text,
+    // }))
+    // }
```
Comment (jurajmajerik) [inline src/extensions/surveys/action-matcher.ts:257]:
```suggestion
```

File: src/utils/survey-event-receiver.ts [hunk 14]
```diff
@@ -43,33 +121,40 @@ export class SurveyEventReceiver {
...
 
     getSurveys(): string[] {
-        const existingActivatedSurveys = this.persistence?.props[SURVEYS_ACTIVATED]
+        const existingActivatedSurveys = this.instance?.persistence?.props[SURVEYS_ACTIVATED]
         return existingActivatedSurveys ? existingActivatedSurveys : []
     }
 
-    getEventRegistry(): Map<string, string[]> {
-        return this.eventRegistry
+    geteventToSurveys(): Map<string, string[]> {
```
Comment (jurajmajerik) [inline src/utils/survey-event-receiver.ts:153]:
```suggestion
    getEventToSurveys(): Map<string, string[]> {
```

**Comment by jurajmajerik** [comment]
Looking solid, good tests! some clean up items :)

**Comment by Phanatic** [comment]
Broken down into two separate PRs

**Pull request closed without merging**